voxblox/include/voxblox/utils/neighbor_tools.h

```diff
@@ -0,0 +1,181 @@
+    CHECK_NOTNULL(layer);
+    layer_ = layer;
+    voxels_per_side_ = layer_->voxels_per_side();
+    CHECK_NE(voxels_per_side_, 0u);
+  }
+
+  void getNeighborsAndDistances(
+      const BlockIndex& block_index, const VoxelIndex& voxel_index,
+      Connectivity connectivity, AlignedVector<VoxelKey>* neighbors,
+      AlignedVector<float>* distances,
+      AlignedVector<Eigen::Vector3i>* directions) const;
+
+  void getNeighbor(const BlockIndex& block_index, const VoxelIndex& voxel_index,
+                   const SignedIndex& direction,
+                   BlockIndex* neighbor_block_index,
+                   VoxelIndex* neighbor_voxel_index) const;
+
+  SignedIndex getOffsetBetweenVoxels(const BlockIndex& start_block_index,
+                                     const VoxelIndex& start_voxel_index,
+                                     const BlockIndex& end_block_index,
+                                     const VoxelIndex& end_voxel_index) const;
+
+ private:
+  const Layer<VoxelType>* layer_;
+
+  size_t voxels_per_side_;
+};
+
+// Uses 26-connectivity and quasi-Euclidean distances.
+// Directions is the direction that the neighbor voxel lives in. If you
+// need the direction FROM the neighbor voxel TO the current voxel, take
+// negative of the given direction.
+template <typename VoxelType>
+void NeighborTools<VoxelType>::getNeighborsAndDistances(
+    const BlockIndex& block_index, const VoxelIndex& voxel_index,
+    Connectivity connectivity, AlignedVector<VoxelKey>* neighbors,
+    AlignedVector<float>* distances,
+    AlignedVector<SignedIndex>* directions) const {
+  CHECK_NOTNULL(layer_);
+  CHECK_NOTNULL(neighbors);
+  CHECK_NOTNULL(distances);
+  CHECK_NOTNULL(directions);
+
+  static const float kSqrt2 = std::sqrt(2);
+  static const float kSqrt3 = std::sqrt(3);
+
+  neighbors->reserve(connectivity);
+  distances->reserve(connectivity);
+  directions->reserve(connectivity);
+
+  VoxelKey neighbor;
+  Eigen::Vector3i direction;
+  direction.setZero();
+  // Distance 1 set.
+  for (unsigned int i = 0; i < 3; ++i) {
+    for (int j = -1; j <= 1; j += 2) {
+      direction(i) = j;
+      getNeighbor(block_index, voxel_index, direction, &neighbor.first,
+                  &neighbor.second);
+      neighbors->emplace_back(neighbor);
+      distances->emplace_back(1.0);
+      directions->emplace_back(direction);
+    }
+    direction(i) = 0;
+  }
+  if (connectivity == Connectivity::kEighteen ||
+      connectivity == Connectivity::kTwentySix) {
+    // Distance sqrt(2) set.
+    for (unsigned int i = 0; i < 3; ++i) {
+      unsigned int next_i = (i + 1) % 3;
+      for (int j = -1; j <= 1; j += 2) {
+        direction(i) = j;
+        for (int k = -1; k <= 1; k += 2) {
+          direction(next_i) = k;
+          getNeighbor(block_index, voxel_index, direction, &neighbor.first,
+                      &neighbor.second);
+          neighbors->emplace_back(neighbor);
+          distances->emplace_back(kSqrt2);
```

Comment:
1. Zec's last review asked me to change the int connectivity value to an enum wiht 3 values since there are only 3 valid values. I agree there, I think it makes more sense to impose only proper connectivity option so that the user isn't surprised to get a number different from what they put in.
2. I think this is a lot worse for readability.